zerver/tests/test_send_email.py

```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
-        self.assertEqual(backend.open.call_count, 1)
-        self.assertEqual(backend.connection.noop.call_count, 1)
-
-        # Test the old connection case when it was closed by the server
-        backend.connection.noop.return_value = [404]
-        backend.close.return_value = False
-        initialize_connection(backend)
-        # 2 more calls to open, 1 more call to noop and 1 call to close
-        self.assertEqual(backend.open.call_count, 3)
-        self.assertEqual(backend.connection.noop.call_count, 2)
-        self.assertEqual(backend.close.call_count, 1)
-
-        # Test backoff procedure
-        backend.open.side_effect = OSError
-        with self.assertRaises(OSError):
+        backend = PersistentSMTPEmailBackend()
+        backend.open()
+        self.assertEqual(backend.opened_at, timezone_now())
+
+        with mock.patch.object(
+            backend, "_open", wraps=backend._open
+        ) as _open_call, mock.patch.object(backend, "close", wraps=backend.close) as close_call:
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
+            initialize_connection(backend)
+
+            self.assertEqual(_open_call.call_count, 1)
+            self.assertEqual(backend.connection.noop.call_count, 1)
+
+            # Test the old connection case when it was closed by the server
+            backend.connection.noop.return_value = [404]
+            initialize_connection(backend)
+
+            # 2 more calls to open and 1 call to close
+            self.assertEqual(_open_call.call_count, 3)
+            self.assertEqual(close_call.call_count, 1)
+
+            # Test backoff procedure
+            _open_call.side_effect = OSError
+            with self.assertRaises(OSError):
+                initialize_connection(backend)
+            # 3 more calls to open as we try 3 times before giving up
+            self.assertEqual(_open_call.call_count, 6)
+
+    @time_machine.travel(timezone_now(), tick=False)
+    def test_max_connection_lifetime(self) -> None:
+        backend = PersistentSMTPEmailBackend()
+        backend.open()
+        self.assertEqual(backend.opened_at, timezone_now())
+
+        with mock.patch.object(
+            backend, "_open", wraps=backend._open
+        ) as _open_call, mock.patch.object(backend, "close", wraps=backend.close) as close_call:
+            # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
             initialize_connection(backend)
-        # 3 more calls to open as we try 3 times before giving up
-        self.assertEqual(backend.open.call_count, 6)
+
+            self.assertEqual(close_call.call_count, 1)
+            self.assertEqual(_open_call.call_count, 2)
+            self.assertEqual(backend.opened_at, timezone_now())
```

Comment:
This is mentioned below -- but this is an unusual case which isn't testing what you think.  Specifically, this is the case where despite the max lifetime = 0, but the connection was left open from the last call, which is not generally what we expect. It does call `open` (to check for being open), `close`, and `open` -- but it would also do that same sequence if the connection _started off closed_, which is not correct.  You should add a test for the "happy path" where `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTE = 0` but the connection starts off closed.  It should call one `open` and one `close`.